pkg/balance/fetcher/fetcher.go

```
@@ -0,0 +1,39 @@
+package fetcher
+
+//go:generate mockgen -destination=mock/fetcher.go . RPCClient,BatchCaller,BalanceScanner
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	gethrpc "github.com/ethereum/go-ethereum/rpc"
+)
+
+type RPCClient interface {
+	ChainID(ctx context.Context) (*big.Int, error)
+	BatchCaller
+	bind.ContractCaller
+}
+
+func FetchNativeBalances(
+	ctx context.Context,
+	addresses []common.Address,
+	atBlock gethrpc.BlockNumber,
+	rpcClient RPCClient,
+	batchSize int,
+) (BalancePerAccountAddress, error) {
+	chainID, err := rpcClient.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// For efficiency, try to use balancescanner first if available for this chain
+	if balanceScanner := getBalanceScanner(chainID.Uint64(), atBlock, rpcClient); balanceScanner != nil {
+		return fetchNativeBalancesWithBalanceScanner(ctx, addresses, atBlock, balanceScanner, batchSize)
+	}
+
+	// As last resort, use less efficient batch call
+	return fetchNativeBalancesStandard(ctx, addresses, atBlock, rpcClient, batchSize)
```

I wonder if it should try`fetchNativeBalancesStandard` when `fetchNativeBalancesWithBalanceScanner` returns err?

---

hmm I'm in general against implicitly hiding errors with retries/fallbacks. If there were any consistent reproducible reason why the BalanceScanner should fail and the standard one should succeed, it should be explicitly handled in the logic itself.